File: Reverse_Words_in_a_String.java
/*151. Reverse Words in a String Add to List
Description  Submission  Solutions
Total Accepted: 144940
Total Submissions: 921627
Difficulty: Medium
Contributors: Admin
Given an input string, reverse the string word by word.

For example,
Given s = "the sky is blue",
return "blue is sky the".

Update (2015-02-12):
For C programmers: Try to solve it in-place in O(1) space.

click to show clarification.

Clarification:
What constitutes a word?
A sequence of non-space characters constitutes a word.
Could the input string contain leading or trailing spaces?
Yes. However, your reversed string should not contain leading or trailing spaces.
How about multiple spaces between two words?
Reduce them to a single space in the reversed string.
Hide Company Tags Microsoft Snapchat Apple Bloomberg Yelp
Hide Tags String
Hide Similar Problems (M) Reverse Words in a String II
*/




/////////////////////////////////////////////////////////////////////////////////////
//C++




/*
Solution 3 faster than Solution 2 faster than Solution 1

Solution 1: split and trim, for
1. split the string
2. insert from the end of the string[]

Solution 2: split, trim, collections.reverse
use Collections.reverse(Arrays.asList(str));
then join String.join(" ", str);

Solution 3: Stringbuilder
use while loop to check if current substring is a word
i is the start of the word
end is the end of the word
at the end str.toString().trim()
*/

/////////////////////////////////////////////////////////////////////////////////////
//Java
//Solution 1: split and trim, for
public class Solution {
    public String reverseWords(String s) {
        String[] str = s.trim().split("\\s+"); //split by multiple whitespace
        String ret = ""; //the return string
        for(int i = str.length-1; i > 0; i--){
            ret += str[i] + " "; //use space to split the diff word
        }
        return ret + str[0];
    }
}


//Solution 2: collections.reverse, string.join
public class Solution {
    public String reverseWords(String s) {
        if(s.length()==0) return s;
        String[] str = s.trim().split("\\s+");
        Collections.reverse(Arrays.asList(str));
        return String.join(" ", str);
    }
}



//Solution 3: stringbuilder
public class Solution {
    public String reverseWords(String s) {
        if(s.length() == 0) return s;
        StringBuilder str = new StringBuilder();

        int len = s.length();
        int end = len-1;
        //i is the start of the substring
        for(int i = len-1; i >= 0; --i){
            if(s.charAt(i) == ' '){
                continue;
            }
            end = i;
            while(i>=0 && s.charAt(i)!=' '){
                i--;
            }
            str.append(s.substring(i+1, end+1)).append(" ");
        }
        return str.toString().trim(); //rememeber trim()
    }
}

File: Reverse_Words_in_a_String.cpp
/*
Given an input string s, reverse the order of the words.

A word is defined as a sequence of non-space characters. The words in s will be separated by at least one space.

Return a string of the words in reverse order concatenated by a single space.

Note that s may contain leading or trailing spaces or multiple spaces between two words. The returned string should only have a single space separating the words. Do not include any extra spaces.



Example 1:

Input: s = "the sky is blue"
Output: "blue is sky the"
Example 2:

Input: s = "  hello world  "
Output: "world hello"
Explanation: Your reversed string should not contain leading or trailing spaces.
Example 3:

Input: s = "a good   example"
Output: "example good a"
Explanation: You need to reduce multiple spaces between two words to a single space in the reversed string.
Example 4:

Input: s = "  Bob    Loves  Alice   "
Output: "Alice Loves Bob"
Example 5:

Input: s = "Alice does not even like bob"
Output: "bob like even not does Alice"


Constraints:

1 <= s.length <= 104
s contains English letters (upper-case and lower-case), digits, and spaces ' '.
There is at least one word in s.


Follow up:

Could you solve it in-place with O(1) extra space?
*/

/*
Solution 1
use find and + to split and join the string

Time Complexity: O(n)
Space Complexity: O(1)
*/
class Solution {
public:
    string reverseWords(string s) {
        // split
        vector<string> splitted;
        size_t pos;
        while( (pos = s.find(' ')) != string::npos) {
            string sub = s.substr(0, pos);
            if (sub.size() >= 1) {
                splitted.push_back(sub);
            }
            s.erase(0, pos+1);
        }
        if (s.size() >= 1) {
            splitted.push_back(s);
        }

        // join
        if (splitted.size() == 0) { return ""; }
        string joined = "";
        for(int i = splitted.size()-1; i >= 0; i--) {
            joined += splitted[i] + ' ';
        }
        // remove last ' '
        joined.erase(joined.size()-1);
        return joined;
    }
};
